src/components/pagesPanel/Pages.tsx

```diff
@@ -0,0 +1,116 @@
+import { useCallback, useEffect, useState } from 'react';
+import { Page } from '@chili-publish/studio-sdk';
+import {
+    PreviewCard,
+    PreviewCardVariant,
+    PreviewType,
+    useMobileSize,
+    useTheme,
+} from '@chili-publish/grafx-shared-components';
+import { ScrollableContainer, Card, Container } from './Pages.styles';
+import { PREVIEW_FALLBACK } from '../../utils/constants';
+import { PageSnapshot } from '../../types/types';
+import { PreviewCardBadge } from './PreviewCardBadge';
+
+interface PagesProps {
+    pages: Page[];
+    activePageId: string | null;
+    pagesToRefresh: string[];
+    setPagesToRefresh: React.Dispatch<React.SetStateAction<string[]>>;
+}
+
+function Pages({ pages, activePageId, pagesToRefresh, setPagesToRefresh }: PagesProps) {
+    const theme = useTheme();
+    const [pageSnapshots, setPageSnapshots] = useState<PageSnapshot[]>([]);
+    const isMobileSize = useMobileSize();
+
+    const handleSelectPage = async (pageId: string) => {
+        await window.StudioUISDK.page.select(pageId);
+    };
+
+    const getPagesSnapshot = useCallback(async (p: Page[]) => {
+        const snapArray = p.map((page) =>
+            window.SDK.page.getSnapshot(page.id).then((snapshot) => ({
+                id: page.id,
+                snapshot: snapshot as unknown as Uint8Array,
+            })),
+        );
+        const awaitedArray = await Promise.all(snapArray);
+        setPageSnapshots(awaitedArray);
+    }, []);
+
+    useEffect(() => {
+        if (pages?.length && !pageSnapshots.length) {
+            getPagesSnapshot(pages);
+        }
+    }, [pages, pageSnapshots, getPagesSnapshot]);
+
+    useEffect(() => {
+        if (!pagesToRefresh.length) return;
+
+        const updateSnapshots = async () => {
+            const updatedSnapshots = await Promise.all(
+                pagesToRefresh.map(async (pageId) => {
+                    const newSnap = (await window.SDK.page.getSnapshot(pageId)) as unknown as Uint8Array;
```

Comment:
won't be executed in parallel. You have to do similar mapping form as in line 33